cfgmgr/coppmgr.cpp

```diff
@@ -0,0 +1,859 @@
+{
+    for (auto it: trap_feature_map)
+    {
+        if (it.second ==  feature)
+        {
+            if (enable)
+            {
+                if (m_coppTrapDisabledMap.find(it.first) != m_coppTrapDisabledMap.end())
+                {
+                    if (m_coppTrapIdTrapGroupMap.find(it.first) != m_coppTrapIdTrapGroupMap.end())
+                    {
+                        string trap_group = m_coppTrapIdTrapGroupMap[it.first];
+
+                        /* Enable only when the restricted group is already pending */
+                        if ((m_coppGroupRestrictedMap.find(trap_group) 
+                             != m_coppGroupRestrictedMap.end()) &&
+                             (checkIfTrapGroupFeaturePending(trap_group)))
+                        {
+                            vector<FieldValueTuple> fvs;
+                            vector<FieldValueTuple> modified_fvs;
+                            string trap_ids;
+
+                            for (auto i : m_coppGroupRestrictedMap[trap_group])
+                            {
+                                FieldValueTuple fv(i.first, i.second);
+                                fvs.push_back(fv);
+                            }
+                            getTrapGroupTrapIds(trap_group, trap_ids);
+                            if (!trap_ids.empty())
+                            {
+                                FieldValueTuple fv(COPP_TRAP_ID_LIST_FIELD, trap_ids);
+                                fvs.push_back(fv);
+                            }
+                            coppGroupGetModifiedFvs (trap_group, fvs, modified_fvs, false);
+                            if (!modified_fvs.empty())
+                            {
+                                m_appCoppTable.set(trap_group, modified_fvs);
+                            }
+                            setCoppGroupStateOk(trap_group);
+                        }
+                    }
+                    m_coppTrapDisabledMap.erase(it.first);
+                }
+            }
+            else
+            {
+                if (m_coppTrapDisabledMap.find(it.first) == m_coppTrapDisabledMap.end())
+                {
+                    m_coppTrapDisabledMap[it.first] = true;
+                    if (m_coppTrapIdTrapGroupMap.find(it.first) != m_coppTrapIdTrapGroupMap.end())
+                    {
+                        string trap_group = m_coppTrapIdTrapGroupMap[it.first];
+
+                        /* Delete the restricted copp group when it goes to pending state
+                         * on feature disable
+                         */
+                        if ((m_coppGroupRestrictedMap.find(trap_group) 
+                             != m_coppGroupRestrictedMap.end()) &&
+                             (checkIfTrapGroupFeaturePending(trap_group)))
+                        {
+                            m_appCoppTable.del(trap_group);
+                            delCoppGroupStateOk(trap_group);
```

Comment:
This changes just the state DB right, should we update state DB only after successfully process the COPP for the feature.

Reply:
Yes. The statedb reflects whether a trap has been installed or not. Sometimes the trap might be present but might not have installed due to the feature being disabled. The statedb will therefore will not contain the state for the trap. Same applies to trap group.